Feature 4863/timber upgrade

mikehardy opened the pull request:
## Purpose / Description

Our Timber dependency was very old, and newer ones allow support for logging during Unit tests, which would be useful

## Fixes
#4863 

## Approach
Between direct observation and conversation here, assumed three requirements for the production logger 
- verbose and debug should be ignored
- info gets a fixed log tag
- warn+error get fixed tag + logging classname prepended to message

I then took the default implementation proposed by the current Timber release documentation, as well as the DebugTree classname discovery implementation, and implemented it here to match the three requirements

## How Has This Been Tested?

I looked at live logcat snippets of production releases and debug releases, then made sure with visual inspection they looked similar.

I also implemented a unit test to exercise the new production logger and verify the three requirements were met

## Learning (optional, can help others)

PowerMockito for static classes is pretty interesting. Here's how to mock statics and verify:
https://automationrhapsody.com/verify-static-method-called-powermock/

Here's the example Timber implementation:
https://github.com/JakeWharton/timber/blob/master/timber-sample/src/main/java/com/example/timber/ExampleApp.java

Looking at our logging implementation prior to any changes is useful 
https://github.com/ankidroid/Anki-Android/blob/v2.5.4/AnkiDroid/src/main/java/com/ichi2/anki/AnkiDroidApp.java#L351

mikehardy commented:
Oh no! It worked locally but failed in CI. I'll have to work out why, but in the meantime you can review the implementation - it shouldn't change much

timrae commented on AnkiDroid/src/main/java/com/ichi2/anki/AnkiDroidApp.java:
```diff
@@ -381,66 +383,35 @@ private static boolean isCurrentLanguage(String l) {
+                    break;
+
+                case Log.WARN:
+                    Log.w(AnkiDroidApp.TAG, message, t);
```
I think the idea here in the current implementation was that I wanted the warnings and errors to follow the same style as the debug logger which contains extra stack trace information I believe...? I think I just copy pasted the necessary code from the upstream debug tree implementation, which is here:

https://github.com/JakeWharton/timber/blob/master/timber/src/main/java/timber/log/Timber.java#L565

timrae replied:
Oh or maybe it's just the name of the class that's generating the message

mikehardy replied:
Hmm - the intention is to be drop-in and avoid behavior change. I am test-verifying regular messages to make sure they have are unchanged as well as visually inspecting (and I've implemented that) but I haven't looked at the exception logging and that's a big oversight your comment clued me in on. I'll do so and try to get it lined up. Thanks for the review

timrae replied:
My memory is a bit hazy, but from looking at the code I think you'll find that in the current production logger, warnings and errors are using the name of the class as the tag instead of `AnkiDroidApp.TAG`.

mikehardy commented:
I think this is ready again after your pointer to the 3rd requirement (warn+error get simple class name prefix in message).

I added a bit more comment about the requirements in the code so future people will have the blueprint

- 2 apologies,
-- I reformatted the code according to checkstyle (codestyle PR related) and the big array part was painful to revert - reverted the rest and that's just in a contiguous block at least
-- I squashed the functional commit and the test commit into one - when reworking I had to touch both function and test and the changes became co-mingled

timrae commented:
Can you use sed to get rid of the formatting change?

mikehardy commented:
I could -check the link I just edited into the last comment - if the split/ignore view doesn't make it clear enough I'll whittle it down

timrae commented:
It's clear enough, I just don't want to commit it to the git history, and it seems trivial to fix.

mikehardy commented:
got it - glad the logic seems okay (pending merge or more discussion) and I'll minimize the change and resubmit after with an explicit request for another look

Completely separate but related - in the future, what's the best way to submit a whitespace change while working on a PR? maybe
- checkout master && import current checkstyle
- reformat file with no other changes and commit
- submit non-whitespace changes in coherent commits?

The motivation for asking is that I'd never alter whitespace normally but it's possible to maybe make it conform file by file if there was convenient and compliant way to submit it

timrae commented:
> * checkout master && import current checkstyle

Well first of all the kind of "obvious part": 
In principle every new PR starts with the process:

```
git checkout master
git pull upstream master
git checkout -b my-feature-branch
```

and ideally every time something is committed upstream between creating the feature branch and merging it, the following will also be done:

```
git checkout master
git pull upstream master
git checkout -
git rebase master
```

Doing this ensures that every PR contains only "fast forward commits", and therefore that what is run on the CI is the exact same as the state the master branch would be in after a `git merge`. In general I only ask contributors to rebase if I think it's necessary for some reason, but in principle you should always do it.

> * reformat file with no other changes and commit
> * submit non-whitespace changes in coherent commits?

Dealing with formatting is a giant pain in the butt. The best way to do it is with a large scale bulk reformatting commit, which is what I'll eventually do, but not until this browser stuff is merged as I don't want to cause unnecessary pain there.

Until then, yeah the best way is probably to either temporarily disable auto-reformatting, or to do an auto-reformatting as a separate commit when you first create the feature branch to make sure it doesn't get muddled up with your code changes.

mikehardy commented:
@timrae  Just re-pushed with no formatting co-mingled

mikehardy commented:
I just realized I left the fixmes in there from before I implemened the tagging, needs one more polish and resubmit

mikehardy commented:
Okay, @timrae no more fixmes dangling around, no whitespace, I think this is actually ready now
